server/consumer.go

```diff
@@ -1031,47 +1051,174 @@ func (o *Consumer) needAck(sseq uint64) bool {
+type waitingRequest struct {
+	client  *client
+	reply   string
+	n       int // For batching
+	expires time.Time
+	noWait  bool
+}
+
+// waiting queue for requests that are waiting for new messages to arrive.
+type waitQueue struct {
+	rp, wp int
+	reqs   []*waitingRequest
+}
+
+// Create a new ring buffer with at most max items.
+func newWaitQueue(max int) *waitQueue {
+	return &waitQueue{rp: -1, reqs: make([]*waitingRequest, max)}
+}
+
+var (
+	errWaitQueueFull = errors.New("wait queue is full")
+	errWaitQueueNil  = errors.New("wait queue is nil")
+)
+
+// Adds in a new request.
+func (wq *waitQueue) add(req *waitingRequest) error {
+	if wq == nil {
+		return errWaitQueueNil
+	}
+	if wq.isFull() {
+		return errWaitQueueFull
+	}
+	wq.reqs[wq.wp] = req
+	// TODO(dlc) - Could make pow2 and get rid of mod.
+	wq.wp = (wq.wp + 1) % cap(wq.reqs)
+
+	// Adjust read pointer if we were empty.
+	if wq.rp < 0 {
+		wq.rp = 0
+	}
+
+	return nil
+}
+
+func (wq *waitQueue) isFull() bool {
+	return wq.rp == wq.wp
+}
+
+func (wq *waitQueue) len() int {
+	if wq == nil || wq.rp < 0 {
+		return 0
+	}
+	if wq.rp < wq.wp {
+		return wq.wp - wq.rp
+	}
+	return cap(wq.reqs) - wq.rp + wq.wp
+}
+
+// Peek will return the next request waiting or nil if empty.
+func (wq *waitQueue) peek() *waitingRequest {
+	if wq == nil {
+		return nil
+	}
+	var wr *waitingRequest
+	if wq.rp >= 0 {
+		wr = wq.reqs[wq.rp]
+	}
+	return wr
+}
+
+// pop will return the next request and move the read cursor.
+func (wq *waitQueue) pop() *waitingRequest {
+	wr := wq.peek()
+	if wr != nil {
+		wr.n--
+		if wr.n <= 0 {
+			wq.reqs[wq.rp] = nil
+			wq.rp = (wq.rp + 1) % cap(wq.reqs)
+			// Check if we are empty.
+			if wq.rp == wq.wp {
+				wq.rp, wq.wp = -1, 0
+			}
 		}
 	}
-	return bs
+	return wr
 }
 
 // processNextMsgReq will process a request for the next message available. A nil message payload means deliver
 // a single message. If the payload is a number parseable with Atoi(), then we will send a batch of messages without
 // requiring another request to this endpoint, or an ACK.
-func (o *Consumer) processNextMsgReq(_ *subscription, _ *client, _, reply string, msg []byte) {
-	// Check payload here to see if they sent in batch size.
-	batchSize := batchSizeFromMsg(msg)
-
+func (o *Consumer) processNextMsgReq(_ *subscription, c *client, _, reply string, msg []byte) {
 	o.mu.Lock()
 	mset := o.mset
 	if mset == nil || o.isPushMode() {
 		o.mu.Unlock()
 		return
 	}
-	shouldSignal := false
+
+	sendErr := func(status int, description string) {
+		sendq := mset.sendq
+		o.mu.Unlock()
+		hdr := []byte(fmt.Sprintf("NATS/1.0 %d %s\r\n\r\n", status, description))
+		pmsg := &jsPubMsg{reply, reply, _EMPTY_, hdr, nil, nil, 0}
+		sendq <- pmsg // Send message.
+	}
+
+	if o.waiting.isFull() {
+		// If our waiting queue is full return an empty response with the proper header.
+		// FIXME(dlc) - Should we do advisory here as well?
+		sendErr(500, "WaitQueue Exceeded")
```

Comment:
So when a user gets to this situation what can he do to resolve this error? Basically nothing right? Would it not be better to before sending this cycle through and check which of the waiting requests have no interest and discard them - or maybe just shift the oldest one off? 

Worried about the UX of this going full and users just being stuck with nothing they can do, some rogue/badly behaving consumer can DOS this easily this way.

Also I think this would be something people will want to change at runtime so perhaps the consumer max waiting would end up being something a consumer udpate can edit.

But as per slack I am not sure I fully grasp what the use case you were aiming at here so lets catch up and make sure I understand what this is intended to achieve

Reply:
You are correct, needed to add one more test to take care of that and try to purge requests before failing. Will add that test and change now.

Reply:
After our call I agree with you we should pop oldest and signal and never return a requests full error. Stabilizes the clients much better, otherwise the later ones would spin on the full error.